src/libsyntax/parse/parser.rs

```diff
@@ -4172,9 +4172,7 @@ impl<'a> Parser<'a> {
                 );
                 err.emit();
                 self.bump();
-            } else if self.eat(&token::BinOp(token::Or)) {
-                // No op.
-            } else {
+            } else if !self.eat(&token::BinOp(token::Or)) {
```

Comment:
I'm not in favour of this. Eating a token very much seems like an "action", rather than just testing a condition, so I think it slightly obfuscates the logic to negate it. Let's leave it how it is.